### ISSUE TITLE
Addition of Adaptive pooling

### ISSUE THREAD

**Issue by kartikdutt18** [mlpack/mlpack#2201]
Hi everyone, I have implemented adaptive pooling and sent a PR for the same however there are certain aspects that I'm not sure about. The layer is very similar to existing pooling layers but every deep learning library had a similar so I added it but as @saksham189 pointed out there is code duplicity to certain extent. So I have summarized the conversation below and I wanted to know how I should go about implementing it.

Why I implemented it as layer?

1. Constructor with only 2 values (output shape)
   **Simpler constructor**, No need for a string to show adaptive padding etc.
  I can't overload constructor in pooling as only 2 values there also don't have default value.
  So a   user would have to do something really different to what they have done in Pytorch.
  Eg. I can't create a constructor saying MaxPooling(size_t outputWidth, size_t outputHeight).

2. **Develop Similarity to Pytorch/tensorflow/keras.**
  It would be easier to use to translate code in Pytorch that uses such layers rather than calculating kernel width and sizes. 

3. **Data Encapsulation.** If say, we added adapting pooling in existing pooling layers, a user shouldn't be able to change kernel width / height. i.e. There should be no function `AdaptivePooling <> module1(2,2).KernelWidth();`

4. **Even in Pytorch the code base is very similar**, If needed I can create objects for max pooling and mean pooling and use them internally as Pytorch does but I didn't want to do that since in adaptive pooling padding is always zero and pooling layer will eventually change. 

5. **Increase user accessibility.**

6. **When pooling layers get padding**, their pooling operations will undergo changes however mine will still remain the same as padding will always be zero.

Cons Of Implementing it as a layer.

1. As pointed out by @saksham189, **Code duplicity** (pooling operations are similar).
   I can try reducing this by making changes in pooling options but codebase will still be similar  since they are similar layers and mlpack follows a template.

Implementing it as overloaded constructor.

1. I thought another neat way of doing this would be by adding a constructor in such a way
    MaxPooling(size_t outputWidth, size_t outputHeight, string adaptiveType)
    With support to input a tuple as well.

Implementing it as wrapper over Pooling Layers.

1. I, personally wanted to avoid that as pooling will eventually get padding and then there will be difference in code base. 
All suggestions are welcome, I wanted to implement it such a way that is more acceptable with the community. Feel free to share your views. 
Also I am sorry, if we aren't supposed to open issues for this but there many ways of implementing the same and I would love learn the different ways one can go about doing the same.
Thanks.

**Comment by kartikdutt18** [comment]
Hi @zoq, what do you think on how it should be implemented?
Thanks.

**Comment by saksham189** [comment]
I think it would be fine to just have the conversation at one place. In this case on your PR https://github.com/mlpack/mlpack/pull/2195 . Otherwise it gets difficult to keep track of the discussion.

**Comment by kartikdutt18** [comment]
Hi @saksham189, so I am closing this then.
